src/Microsoft.AspNet.Cors.Core/CorsResult.cs

```diff
@@ -0,0 +1,179 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Text;
+
+namespace Microsoft.AspNet.Cors.Core
+{
+    /// <summary>
+    /// Results returned by <see cref="ICorsEngine"/>.
+    /// </summary>
+    public class CorsResult
+    {
+        private long? _preflightMaxAge;
+
+        /// <summary>
+        /// Gets a value indicating whether the result is valid.
+        /// </summary>
+        public bool IsValid
+        {
+            get
+            {
+                return ErrorMessages.Count == 0;
+            }
+        }
+
+        /// <summary>
+        /// Gets the error messages.
+        /// </summary>
+        public IList<string> ErrorMessages { get; } = new List<string>();
+
+        /// <summary>
+        /// Gets or sets the allowed origin.
+        /// </summary>
+        public string AllowedOrigin { get; set; }
+
+        /// <summary>
+        /// Gets or sets a value indicating whether the resource supports user credentials.
+        /// </summary>
+        public bool SupportsCredentials { get; set; }
+
+        /// <summary>
+        /// Gets the allowed methods.
+        /// </summary>
+        public IList<string> AllowedMethods { get; } = new List<string>();
+
+        /// <summary>
+        /// Gets the allowed headers.
+        /// </summary>
+        public IList<string> AllowedHeaders { get; } = new List<string>();
+
+        /// <summary>
+        /// Gets the allowed headers that can be exposed on the response.
+        /// </summary>
+        public IList<string> AllowedExposedHeaders { get; } = new List<string>();
+
+        /// <summary>
+        /// Gets or sets the number of seconds the results of a preflight request can be cached.
+        /// </summary>
+        public long? PreflightMaxAge
+        {
+            get
+            {
+                return _preflightMaxAge;
+            }
+            set
+            {
+                if (value < 0)
+                {
+                    throw new ArgumentOutOfRangeException("value", Resources.PreflightMaxAgeOutOfRange);
+                }
+                _preflightMaxAge = value;
+            }
+        }
+
+        /// <summary>
+        /// Returns CORS-specific headers that should be added to the response.
+        /// </summary>
+        /// <returns>The response headers.</returns>
+        public virtual IDictionary<string, string> ToResponseHeaders()
```

Comment:
Move this logic to `ICorsEngine.ApplyResult(HttpResponse, CorsResult)` method - data objects should avoid having behavior.